sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+        case j @ Join(_, _, _: InnerLike, _) if !j.ordered =>
+          reorder(j, j.outputSet)
+      }
+    }
+  }
+
+  def reorder(plan: LogicalPlan, output: AttributeSet): LogicalPlan = {
+    val (items, conditions) = extractInnerJoins(plan)
+    val result =
+      if (items.size > 2 && items.size <= conf.joinReorderDPThreshold && conditions.nonEmpty) {
+        JoinReorderDP(conf, items, conditions, output).search().getOrElse(plan)
+      } else {
+        plan
+      }
+    // Set all inside joins ordered.
+    setOrdered(result)
+    result
+  }
+
+  /**
+   * Extract inner joinable items and join conditions.
+   * This method works for bushy trees and left/right deep trees.
+   */
+  def extractInnerJoins(plan: LogicalPlan): (Seq[LogicalPlan], Set[Expression]) = plan match {
+    case j @ Join(left, right, _: InnerLike, cond) =>
+      val (leftPlans, leftConditions) = extractInnerJoins(left)
+      val (rightPlans, rightConditions) = extractInnerJoins(right)
+      (leftPlans ++ rightPlans, cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++
+        leftConditions ++ rightConditions)
+    case Project(_, j @ Join(left, right, _: InnerLike, cond)) =>
+      val (leftPlans, leftConditions) = extractInnerJoins(left)
+      val (rightPlans, rightConditions) = extractInnerJoins(right)
+      (leftPlans ++ rightPlans, cond.map(splitConjunctivePredicates).getOrElse(Nil).toSet ++
+        leftConditions ++ rightConditions)
+    case _ =>
+      (Seq(plan), Set())
+  }
+
+  def setOrdered(plan: LogicalPlan): Unit = plan match {
+    case j @ Join(left, right, _: InnerLike, cond) =>
+      j.ordered = true
+      setOrdered(left)
+      setOrdered(right)
+    case Project(_, j @ Join(left, right, _: InnerLike, cond)) =>
+      j.ordered = true
+      setOrdered(left)
+      setOrdered(right)
+    case _ =>
+  }
+}
+
+/**
+ * Reorder the joins using a dynamic programming algorithm:
+ * First we put all items (basic joined nodes) into level 1, then we build all two-way joins
+ * at level 2 from plans at level 1 (single items), then build all 3-way joins from plans
+ * at previous levels (two-way joins and single items), then 4-way joins ... etc, until we
+ * build all n-way joins and pick the best plan among them.
+ *
+ * When building m-way joins, we only keep the best plan (with the lowest cost) for the same set
+ * of m items. E.g., for 3-way joins, we keep only the best plan for items {A, B, C} among
+ * plans (A J B) J C, (A J C) J B and (B J C) J A.
+ *
+ * Thus the plans maintained for each level when reordering four items A, B, C, D are as follows:
+ * level 1: p({A}), p({B}), p({C}), p({D})
+ * level 2: p({A, B}), p({A, C}), p({A, D}), p({B, C}), p({B, D}), p({C, D})
+ * level 3: p({A, B, C}), p({A, B, D}), p({A, C, D}), p({B, C, D})
+ * level 4: p({A, B, C, D})
+ * where p({A, B, C, D}) is the final output plan.
+ *
+ * For cost evaluation, since physical costs for operators are not available currently, we use
```

Comment:
This is a good point. Ideally, choices of join order, join algorithm and enforcer like distribute and sort are made in the same algorithm, but two issues make this difficult to do in Spark.
1. Catalyst separates logical optimization, physical implementation (strategies) and enforcer (EnsureRequirements) into different stages.
2. If we want to consider choosing algorithms and enforcers, we need physical cost evaluation to guide us, while we don't have this now.
So in the current stage, join reordering is logical optimization. In the future after we support physical cost evaluation and maybe modify catalyst's structure, we can have a more mature algorithm. The DP framework will still work, the difference should be the logic of plan cost comparison.